### PR TITLE
v0.1.3

### PR DESCRIPTION
**Pull request by HollowMan6**
- Add support for Blockly v9:
  - `is3dSupported` is removed, so now make `Blockly.WorkspaceSvg.getBlockDragSurface` always return `null` to disable dragsurface: https://github.com/google/blockly/pull/6400
  - Use Blockly.getSelected() as Blockly.selected is now deprecated: https://github.com/google/blockly/blob/52a0d525d74fa88ba3154a60feb3014034552377/core/main.js#L132-L133
- Add back dragselect dependency as now the PR has got merged and released: https://github.com/ThibaultJanBeyer/DragSelect/releases/tag/v2.4.4

Signed-off-by: Hollow Man <hollowman@opensuse.org>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -86,7 +86,7 @@ multiselectPlugin.init(options);
 - `inMultipleSelectionModeWeakMap`: The WeakMap storing whether the plugin is in multiple selection mode by workspace svg.
 
 ## Credit
-- [ds.min.js](lib/ds.min.js): This plugin uses [a patched version of DragSelect](https://github.com/ThibaultJanBeyer/DragSelect/pull/128) to realize "drag a rectangle to select multiple blocks" feature.
+- [DragSelect]([lib/ds.min.js](https://github.com/ThibaultJanBeyer/DragSelect)): This plugin uses [a patched version of DragSelect](https://github.com/ThibaultJanBeyer/DragSelect/pull/143) to realize "drag a rectangle to select multiple blocks" feature (has got merged into v2.4.4).
```
Comment (mark-friedman) [inline README.md:89]: It looks like you are no longer using your own patched version, so you can remove references to `lib/ds.min.js` and can remove the references to the patched version and the pull request.

**Review by mark-friedman — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (HollowMan6) [reply README.md]: Sorry about this, it was just some stupid VSCode auto-completion that I didn't notice...

**Comment by mark-friedman** [comment]
>Sorry about this, it was just some stupid VSCode auto-completion that I didn't notice...

No problem!  It looks the Readme still mentions using a patched version of the library, but I'm think that's no longer the case, since your PR was merged into DragSelect.

**Comment by HollowMan6** [comment]
> No problem! It looks the Readme still mentions using a patched version of the library, but I'm think that's no longer the case, since your PR was merged into DragSelect.

Ah Okay, has fixed this.

**Review by mark-friedman — APPROVED** [review]
I did some testing and all seem fine.  I noticed a small nit, which is that in a newly refreshed/reloaded workspace the very first time I shift-click a child block it only selects the child.  Once I click away, from that point on when I shift-click a child block it selects the child and the top-level ancestor block.

**Comment by HollowMan6** [comment]
> I noticed a small nit, which is that in a newly refreshed/reloaded workspace the very first time I shift-click a child block it only selects the child. Once I click away, from that point on when I shift-click a child block it selects the child and the top-level ancestor block.

Hi, thanks for testing! Actually I can't reproduce what you mentioned. In a newly refreshed/reloaded workspace the very first time I shift-click a child block it only selects the child, then when I click away, from that point on when I shift-click that child block again, it will still select that child block only.

GIF 2022-11-12 9-30-33

Maybe I have misunderstood what you said, it would be great if you can post some gifs to help me know.

**Comment by ewpatton** [comment]
@mark-friedman I'm also not seeing the behavior as described. So far things seem to be working well for me.

**Comment by mark-friedman** [comment]
@HollowMan6 @ewpatton 
Below is a video of what I'm seeing.  I have a clone of HollowMan6/multi-select and have checked out the main branch. I've run `npm install` and I'm running the Blockly Playground in that directory via `npm start`.  I pulled out the blocks that you see and then did a browser reload.  This is on a Mac and I see the same behavior on Chrome Version 107.0.5304.110, Firefox Version 106.0.2 and Safari Version 16.0

Note that after the reload my first click is the shift-click of the `print` block.  All the clicks on the block are shift-clicks.  After each shift-click there is a regular click on the workspace background.

https://user-images.githubusercontent.com/1743690/202016808-c6a443d5-f531-43e9-be17-342ccd7fb547.mov

**Comment by HollowMan6** [comment]
Okay, I see. Unfortunately, I cannot come up with an instant fix for this because this relates to the fundamental of how this plugin works, and I would say that the bug was not introduced by this PR and should exist since the first beginning. Currently, we rely on DragSelect to know which block gets selected. DragSelect seems to listen to the "blocks". However, it actually works by listening to the SVG path element, which is always a rectangle with some transparent parts forming a block.

For such irregularly shaped blocks in your example, the two blocks' SVG path elements overlap with each other, so if you shift-click on that area, DragSelect will think that both blocks get selected.

image
image

I would say a proper fix should be that Blockly implements some sort of "native version of DragSelect" and exposes it with an API, so that we can know for sure where the block actually locates, and that would be a lot of work. If such an API is available, we can then use that API instead of DragSelect.

**Comment by mark-friedman** [comment]
Any idea why it works one way in a freshly loaded workspace and another way after click activity in the background?

**Comment by HollowMan6** [comment]
> Any idea why it works one way in a freshly loaded workspace and another way after click activity in the background?

In a freshly loaded workspace, your focus is not on the workspace until you click on some part of it. The plugin cannot receive the keyboard event when your focus is not on the workspace, so the shift drag didn't actually get activated at that time (You can see that the icon in the right-bottom corner also suggests this). The first click on a freshly loaded workspace would always be the Blockly native one that selects only one block if you happen to click on some block, and that matches the typically expected behaviour as you will not select more blocks in a freshly loaded workspace.

**Comment by ewpatton** [comment]
It's possible we could attach the key handler to the document, or some other developer provided element.

**Comment by HollowMan6** [comment]
> It's possible we could attach the key handler to the document, or some other developer provided element.

I remember we discussed this issue before in the summer, as there can be multiple workspaces with multiple instances of plugins on one page. We can't just simply listen to the document.

After all, the behavior is expected, and I think this should not be a problem to keep as it is.

**Comment by ewpatton** [comment]
That supports my position that it could be developer provided as the developer knows more about their end product than we do. We might also be able at least in app inventor to focus the workspace to address this issue if that's the limitation. It ought to be as easy as recommending in the readme that people call .focus() on the workspace but I have tried that yet.

**Comment by HollowMan6** [comment]
> That supports my position that it could be developer provided as the developer knows more about their end product than we do. We might also be able at least in app inventor to focus the workspace to address this issue if that's the limitation. It ought to be as easy as recommending in the readme that people call .focus() on the workspace but I have tried that yet.

Yeah, I agree with the later part, we can call .focus() to get workspace selected in the first place, and just added that together with the issue discussed above into the README. I've also added the .focus() into the test, now it should not be a limitation any more when running the Blockly Playground via npm start.

For the former part, I think it's not necessary, it may complex the plugin by requiring more parameters. There may also exist some hidden issues if we implement that.

File: package.json [hunk 2]
```diff
@@ -38,13 +38,16 @@
...
-    "@blockly/dev-scripts": "^1.2.26",
-    "@blockly/dev-tools": "^4.0.2",
-    "blockly": "^8.0.4"
+    "@blockly/dev-scripts": "^1.2.28",
+    "@blockly/dev-tools": "^5.0.2",
+    "blockly": "^9.1.0"
   },
   "peerDependencies": {
-    "blockly": "^8.0.4"
+    "blockly": "^9.1.0"
```
Comment (ewpatton) [inline package.json:50]: Do we need this to start at 9.1.0, or can we be more inclusive and start at 9.0.0?

Comment (HollowMan6) [reply package.json]: Of course we can, just updated that.

**Comment by HollowMan6** [comment]
> We should file an issue to look into the DragSelect with invisible rectangles thing. I think it's solveable if we play with the event handling a bit since a child block is contained within its parent in the SVG, so we can always just pick the node deepest in the DOM tree.

Yeah, but that's not a fix that can solve the fundamental issue:
GIF 2022-11-16 19-25-13

A proper fix should still be that Blockly implements some sort of "native version of DragSelect" and exposes it with an API, so that we can know for sure where the block actually locates.

**Comment by HollowMan6** [comment]
Has filed an issue here: https://github.com/ThibaultJanBeyer/DragSelect/issues/150

**Pull request merged**